api/types/access_list.go

```diff
@@ -0,0 +1,274 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package types
+
+import (
+	"time"
+
+	"github.com/gravitational/trace"
+
+	accesslistpb "github.com/gravitational/teleport/api/gen/proto/go/accesslist/v1"
+	"github.com/gravitational/teleport/api/types/header"
+	"github.com/gravitational/teleport/api/types/traits"
+)
+
+// FromAccessListV1 converts a v1 access list into an internal access list object.
+func FromAccessListV1(msg *accesslistpb.AccessList) (*AccessList, error) {
+	owners := make([]*AccessListOwner, len(msg.Spec.Owners))
+	for i, owner := range msg.Spec.Owners {
+		owners[i] = &AccessListOwner{
+			Name:        owner.Name,
+			Description: owner.Description,
+		}
+	}
+
+	members := make([]*AccessListMember, len(msg.Spec.Members))
+	for i, member := range msg.Spec.Members {
+		members[i] = &AccessListMember{
+			Name:    member.Name,
+			Joined:  member.Joined.AsTime(),
+			Expires: member.Expires.AsTime(),
+			Reason:  member.Reason,
+			AddedBy: member.AddedBy,
+		}
+	}
+
+	accessList, err := NewAccessList(header.FromMetadataV1(msg.Header.Metadata), &AccessListSpec{
+		Owners: owners,
+		Audit: &AccessListAudit{
+			Frequency: msg.Spec.Audit.Frequency.AsDuration(),
+		},
+		MembershipRequires: &AccessListRequires{
+			Roles:  msg.Spec.MembershipRequires.Roles,
+			Traits: traits.FromV1(msg.Spec.MembershipRequires.Traits),
+		},
+		OwnershipRequires: &AccessListRequires{
+			Roles:  msg.Spec.OwnershipRequires.Roles,
+			Traits: traits.FromV1(msg.Spec.OwnershipRequires.Traits),
+		},
+		Grants: &AccessListGrants{
+			Roles:  msg.Spec.Grants.Roles,
+			Traits: traits.FromV1(msg.Spec.Grants.Traits),
+		},
+		Members: members,
+	})
+
+	return accessList, trace.Wrap(err)
+}
+
+// AccessList describes the basic building block of access grants, which are
+// similar to access requests but for longer lived permissions that need to be
+// regularly audited.
+type AccessList struct {
+	// ResourceHeader is the common resource header for all resources.
+	*header.ResourceHeader
+
+	// Spec is the specification for the access list.
+	Spec *AccessListSpec `json:"spec" yaml:"spec"`
```

Comment:
Do we typically embed pointers in our resources? This looks unusual to me.

Reply:
It's hard to compare exactly since we're relying so much on gogo, but generated non-gogo proto objects are typically pointers -- I was more or less mimicking this and trying to generally reduce the amount of copy-by-value going on here. I'd argue this would make things slightly more efficient, but I can't imagine the cost would be particularly high regardless.